core/src/main/java/org/dogtagpki/tomcat/JSSNioEndpoint.java

```diff
@@ -0,0 +1,77 @@
+package org.dogtagpki.tomcat;
+
+
+import java.nio.channels.SocketChannel;
+import java.util.List;
+
+import javax.net.ssl.SSLEngine;
+
+import org.apache.juli.logging.Log;
+import org.apache.juli.logging.LogFactory;
+import org.apache.tomcat.util.ExceptionUtils;
+import org.apache.tomcat.util.net.NioChannel;
+import org.apache.tomcat.util.net.NioEndpoint;
+import org.apache.tomcat.util.net.SocketBufferHandler;
+import org.apache.tomcat.util.net.openssl.ciphers.Cipher;
+
+public class JSSNioEndpoint extends NioEndpoint {
+
+    private static final Log log = LogFactory.getLog(NioEndpoint.class);
+    @Override
+    protected boolean setSocketOptions(SocketChannel socket) {
+        NioSocketWrapper socketWrapper = null;
+        try {
+            // Allocate channel and wrapper
+            NioChannel channel = null;
+            if (getNioChannels() != null) {
+                channel = getNioChannels().pop();
+            }
+            if (channel == null) {
+                SocketBufferHandler bufhandler = new SocketBufferHandler(
+                        socketProperties.getAppReadBufSize(),
+                        socketProperties.getAppWriteBufSize(),
+                        socketProperties.getDirectBuffer());
+                if (isSSLEnabled()) {
+                    channel = new JSSSecureNioChannel(bufhandler, this);
```

Comment:
Does the original `SecureNioChannel` have the correct local & remote addresses? I was wondering if we can do something like this:
```
channel = new SecureNioChannel(bufhandler, this);
SecureNioChannel secureNioChannel = (SecureNioChannel) channel;
SSLEngine sslEngine = secureNioChannel.getSslEngine();
JSSSession jssSession = (JSSSession) sslEngine.getSession();
jssSession.setLocalAddr(socketWrapper.getLocalAddr());
jssSession.setRemoteAddr(socketWrapper.getRemoteAddr());
```

Reply:
The SSLEngine is not created with the `SecureNioChannel`  but when the handshake is performed. Actually, when the handshake is requested, among the initial operations there is the SNI process and the SSLEngine is created there: 

https://github.com/apache/tomcat/blob/3e287ac4d7dbc04a364f10ede42b3518866d4ed1/java/org/apache/tomcat/util/net/SecureNioChannel.java#L304

Therefore, you have to perform this operation after the method `performSNI()` and before to complete the method `handshake()` because after the method the audit is written.